Make extract! accept objects with internal hash implementation

bacchir opened the pull request:
In a recent project I've experienced an odd situation of having an object that implements some `Hash` behavior and, even though it has methods such as `[]` and `fetch`, `jbuilder.extract!` did not work with it. So I just created this little PR with some tests to make `jbuilder` more agnostic.

Here's an example (also used in the tests) of the object I'm trying to pass to jbuilder:

```
class PersonWithHash
  attr_reader :name, :collection

  def initialize(name, age)
    @collection = { age: age }
    @name = name
  end

  delegate :[], :fetch, to: :@collection
end
```

robin850 commented on lib/jbuilder.rb:
```diff
@@ -252,12 +250,8 @@ def target!
-  def _extract_method_values(object, attributes)
-    attributes.each{ |key| _set_value key, object.public_send(key) }
+  def _extract_value(object, attribute)
+    object.respond_to?(attribute) ? object.public_send(attribute) : object.fetch(attribute)
```
The risk is pretty low but still, it should rather be:

~~~ruby
object.respond_to?(:fetch) ? object.fetch(attribute) : object.public_send(attribute)
~~~

Otherwise, if a hash or hash-like object has a `:object_id`, `:hash` or any of its method as a key, the wrong value may be returned.

Could you fix this and add a test to ensure that a `Hash` with such key would be properly handled please ?

bacchir replied:
Sorry @robin850 I just saw your comment now. Sure I'll change it 😊

bacchir replied:
@robin850 I changed the implementation the way you suggested, however I changed respond_to?(:fetch) to respond_to?(:[]) to avoid problems when an object have a custom fetch method for other purposes.

bacchir commented:
@robin850 I changed the implementation the way you suggested, however I changed `respond_to?(:fetch)` to `respond_to?(:[])` to avoid problems when an object have a custom `fetch` method for other purposes.

bacchir commented:
@robin850 I changed the implementation the way you suggested, however I changed respond_to?(:fetch) to respond_to?(:[]) to avoid problems when an object have a custom fetch method for other purposes.
